js/packages/openinference-instrumentation-llama-index/src/utils.ts

```diff
@@ -308,3 +388,78 @@ export function patchQueryEmbeddingMethod(
     return context.bind(execContext, wrappedPromise);
   };
 }
+
+export function patchLLMChat(original: LLMChatMethodType, tracer: Tracer) {
+  return function (this: unknown, ...args: Parameters<LLMChatMethodType>) {
+    const span = tracer.startSpan(`llm`, {
+      kind: SpanKind.INTERNAL,
+      attributes: {
+        [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+          OpenInferenceSpanKind.LLM,
+      },
+    });
+
+    const execContext = getExecContext(span);
+
+    const execPromise = safeExecuteInTheMiddle<ReturnType<LLMChatMethodType>>(
+      () => {
+        return context.with(execContext, () => {
+          return original.apply(this, args);
+        });
+      },
+      (error) => handleError(span, error),
+    );
+
+    // Metadata is a property found on the class and not in args
+    span.setAttributes(parseLLMMetadata(this));
+
+    const wrappedPromise = execPromise.then((result) => {
+      span.setAttributes(
+        getLLMChatAttributes({ input: args[0], output: result }),
+      );
+      span.end();
+      return result;
+    });
+    return context.bind(execContext, wrappedPromise);
+  };
+}
+
+export function patchLLMComplete(
+  original: LLMCompleteMethodType,
+  tracer: Tracer,
+) {
+  return function (this: unknown, ...args: Parameters<LLMCompleteMethodType>) {
+    const span = tracer.startSpan(`llm`, {
+      kind: SpanKind.INTERNAL,
+      attributes: {
+        [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+          OpenInferenceSpanKind.LLM,
+      },
+    });
+
+    const execContext = getExecContext(span);
+
+    const execPromise = safeExecuteInTheMiddle<
+      ReturnType<LLMCompleteMethodType>
+    >(
+      () => {
+        return context.with(execContext, () => {
+          return original.apply(this, args);
+        });
+      },
+      (error) => handleError(span, error),
+    );
+
+    // Metadata is a property found on the class and not in args
+    span.setAttributes(parseLLMMetadata(this));
+
+    const wrappedPromise = execPromise.then((result) => {
+      span.setAttributes(
+        getLLMCompleteAttributes({ input: args[0], output: result }),
```

Comment:
we probably need to wrap all of these in safety functions not just stringify, see LangChain instrumentation

Reply:
the ptch functions don't need to be but all of the parsing utils should be